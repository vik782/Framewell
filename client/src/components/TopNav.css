/* hamburger menu */

.nav-icon {
  display: none;
}

.nav-icon-button {
  font-size: 30px;
}

/* navbar styling */

.topnav {
  background: #96b3c2;
  grid-area: nav;
  height: 80px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 0 30px 0 30px;
  box-shadow: 0px 4px 15px rgba(0, 0, 0, 0.2);
  color: white;
  font-weight: 600;
  z-index: 0;
}

.topnav-left {
  display: flex;
  justify-content: space-between;
  align-items: center;
  text-align-last: justify;
}

.topnav-left > .website-logo > img {
  height: 60px;
  margin-right: 10px;
  cursor: pointer;
}

.topnav-right {
  display: flex;
  justify-content: space-between;
  align-items: center;
  text-align-last: justify;
}

.topnav-button > button {
  margin-left: 40px;
  border: none;
  background: none;
  color: white;
  font-family: "Montserrat";
  font-weight: 600;
  cursor: pointer;
  position: relative;
}

.topnav-button > button::before {
  content: "";
  position: absolute;
  bottom: 0;
  right: 0;
  width: 0;
  height: 1px;
  background-color: white;
  transition: width 0.6s cubic-bezier(0.25, 1, 0.5, 1);
}

.spacer {
  margin-bottom: 75px;
}

.logout-icon-button {
  margin-right: 5px;
}

@media only screen and (max-width: 978px) {
  .nav-icon {
    display: inline;
  }

  .topnav-right {
    display: none;
  }
}

@media only screen and (max-width: 480px) {
  .topnav-left {
    display: none;
  }
}

/* Animations */
@media (hover: hover) and (pointer: fine) {
  .topnav-button > button:hover::before {
    left: 0;
    right: auto;
    width: 100%;
    color: white;
  }
}
