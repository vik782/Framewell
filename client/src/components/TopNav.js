/**
 * @fileoverview Abstraction of the main navigation bar (desktop mode)
 * Uses:
 * - React for rendering HTML
 * - Iconify for adding icons
 * - React Router for handling client-side routes
 * - Universal Cookie for handling browser cookies and validating logins
 */

// Imports of packages
import React from "react";
import { Link } from "react-router-dom";
import { Icon } from "@iconify/react";
import Cookies from "universal-cookie";

// Style-based imports
import "./TopNav.css";

// Data-oriented imports
import logo from "../pictures/WebsiteLogo.png";

// obtain token from cookie
const cookies = new Cookies();

/**
 * The Top nav takes in props that decide whether it should show or not
 * and contains links that the user can use to navigate accordingly
 * @return {React.Component}
 */
function TopNav({ mobileNavOpen, openMobileNav }) {
  /**
   * Removes the cookie that authenticated the login and redirects the user
   * back to the login page
   */
  function logout() {
    cookies.remove("TOKEN", { path: "/" });

    // user redirected to login page
    window.location.replace("/login");
  }

  let /** React.Component */ DashboardButton = window.location.href.includes(
      "/dashboard"
    ) ? (
      <div className="topnav-button">
        <button className="dashboard" onClick={() => window.location.reload()}>
          Dashboard
        </button>
      </div>
    ) : (
      /* else */ <Link className="topnav-button" to={`/dashboard`}>
        <button className="dashboard">Dashboard</button>
      </Link>
    );
  let /** React.Component */ WebsiteButton = window.location.href.includes(
      "/dashboard"
    ) ? (
      <div className="website-logo">
        <img
          src={logo}
          alt="Website logo"
          onClick={() => window.location.reload()}
        />
      </div>
    ) : (
      /* else */ <Link className="website-logo" to={`/dashboard`}>
        <img src={logo} alt="website logo" />
      </Link>
    );

  return (
    <nav className="topnav">
      <div className="topnav-left">
        {WebsiteButton}
        Framewell
      </div>
      <div className="topnav-right">
        {DashboardButton}

        <Link className="topnav-button" to={`/add-artefact`}>
          <button className="add-artefact">Add Artefact</button>
        </Link>
        <div className="topnav-button" onClick={() => logout()}>
          <button className="logout">
            <Icon
              className="logout-icon-button"
              icon="icon-park-outline:logout"
            />
            Sign Out
          </button>
        </div>
      </div>

      <div className="nav-icon" onClick={() => openMobileNav()}>
        <Icon className="nav-icon-button" icon="charm:menu-hamburger" />
      </div>
    </nav>
  );
}

export default TopNav;
