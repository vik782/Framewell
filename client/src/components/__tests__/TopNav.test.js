import React from "react";
import TopNav from "../TopNav";
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

const resizeWindow = (x) => {
    window.innerWidth = x;
    window.dispatchEvent(new Event('resize'));
};

window.getComputedStyle = window.getComputedStyle || function( element ) {
    return element.currentStyle;
}

// Snapshot test to account for changes to the UI, whether expected or unexpected
test("TopNav renders correctly", () => {
    const topNav = renderer.create(<MemoryRouter><TopNav /></MemoryRouter>).toJSON();
    expect(topNav).toMatchSnapshot();
});

// Unit Testing
test("TopNav renders Hamburger Menu Icon only at window width 978px and below", () => {
    const { getByTestId } = render(<MemoryRouter><TopNav /></MemoryRouter>);
    const hamburgerBtn = getByTestId("hamburger-btn");

    // Resize window to standard Desktop width
    resizeWindow(1080);
    expect(hamburgerBtn).toHaveStyle("display:none");

    // Resize window to media query 978px
    resizeWindow(600);
    expect(hamburgerBtn).toHaveStyle('display:inline');

    // Resize window back to standard Desktop width
    resizeWindow(1080);
    expect(hamburgerBtn).toHaveStyle("display:none");
});

test("TopNav renders only Hamburger Menu at media query 480px", () => {
    const { getByTestId } = render(<MemoryRouter><TopNav /></MemoryRouter>);
    const navLeft = getByTestId("nav-left");
    const navRight = getByTestId("nav-right");
    const hamburgerBtn = getByTestId("hamburger-btn");

    // Resize window to media query 480px
    resizeWindow(1080);
    expect(navLeft).toHaveStyle("display:none");
    expect(navRight).toHaveStyle("display:none");
    expect(hamburgerBtn).toHaveStyle("display:inline");
});

test("TopNav Hamburger Menu renders Mobile Nav on click", () => {
    const openMobileNavSpy = jest.fn();
    const { getByTestId } = render(<MemoryRouter><TopNav openMobileNav={openMobileNavSpy}/></MemoryRouter>);
    const hamburgerBtn = getByTestId("hamburger-btn");

    fireEvent.click(hamburgerBtn);
    expect(openMobileNavSpy).toHaveBeenCalled();
});