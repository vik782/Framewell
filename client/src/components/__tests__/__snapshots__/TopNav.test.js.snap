// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopNav renders correctly 1`] = `
<nav
  className="topnav"
>
  <div
    className="topnav-left"
    data-testid="nav-left"
  >
    <img
      alt="Website logo"
      src="LS-Logo.png"
    />
    Framewell
  </div>
  <div
    className="topnav-right"
    data-testid="nav-right"
  >
    <a
      className="topnav-button"
      href="/dashboard"
      onClick={[Function]}
    >
      <button
        className="dashboard"
        data-testid="dashboard-btn"
      >
        Dashboard
      </button>
    </a>
    <a
      className="topnav-button"
      href="/add-artefact"
      onClick={[Function]}
    >
      <button
        className="add-artefact"
        data-testid="add-btn"
      >
        Add Artefact
      </button>
    </a>
    <div
      className="topnav-button"
      onClick={[Function]}
    >
      <button
        className="logout"
        data-testid="logout-btn"
      >
        <span />
        Sign Out
      </button>
    </div>
  </div>
  <div
    className="nav-icon"
    data-testid="hamburger-btn"
    onClick={[Function]}
  >
    <span />
  </div>
</nav>
`;
